xcdat/regridder/regrid2.py

```diff
@@ -46,259 +46,199 @@ def __init__(self, input_grid: xr.Dataset, output_grid: xr.Dataset, **options: A
-                    "exists and the attributes are correct."
-                )
+    y_length = len(lat_mapping)
+    x_length = len(lon_mapping)
 
-            if standard_name in self._output_grid:
-                output_sizes[axis_name] = self._output_grid.sizes[standard_name]
-            else:
-                output_sizes[axis_name] = da.sizes[standard_name]
+    other_dims = {
+        x: y for x, y in input_data_var.sizes.items() if x not in (y_name, x_name)
+    }
+    other_sizes = list(other_dims.values())
 
-        return output_sizes
+    data_shape = [y_length * x_length] + other_sizes
+    output_data = np.zeros(data_shape, dtype=np.float32)
 
-    def _regrid(
-        self,
-        input_data: np.ndarray,
-        axis_sizes: Dict[str, int],
-        ordered_axis_names: List[str],
-    ) -> np.ndarray:
-        """Applies regridding to input data.
+    is_2d = input_data_var.ndim <= 2
 
-        Parameters
-        ----------
-        input_data : np.ndarray
-            Input multi-dimensional array on source grid.
-        axis_sizes : Dict[str, int]
-            Mapping of axis name e.g. ("X", "Y", etc) to output sizes.
-        ordered_axis_names : List[str]
-            List of axis name in order of dimensions of ``input_data``.
-
-        Returns
-        -------
-        np.ndarray
-            Multi-dimensional array on destination grid.
-        """
-        input_lat_index = ordered_axis_names.index("Y")
-
-        input_lon_index = ordered_axis_names.index("X")
+    # need to optimize
+    for y in range(y_length):
+        y_seg = np.take(input_data, lat_mapping[y], axis=y_index)
 
-        output_shape = [axis_sizes[x] for x in ordered_axis_names]
+        for x in range(x_length):
+            x_seg = np.take(y_seg, lon_mapping[x], axis=x_index, mode="wrap")
 
-        output_data = np.zeros(output_shape, dtype=np.float32)
+            cell_weight = np.dot(lat_weights[y], lon_weights[x])
 
-        base_put_index = self._base_put_indexes(axis_sizes)
+            output_seg_index = y * x_length + x
 
-        for lat_index, lat_map in enumerate(self._lat_mapping):
-            lat_weight = self._lat_weights[lat_index]
+            if is_2d:
+                output_data[output_seg_index] = np.divide(
+                    np.nansum(
+                        np.multiply(x_seg, cell_weight),
+                        axis=(y_index, x_index),
+                    ),
+                    np.sum(cell_weight),
+                )
+            else:
+                output_seg = output_data[output_seg_index]
 
-            input_lat_segment = np.take(input_data, lat_map, axis=input_lat_index)
+                np.divide(
+                    np.nansum(
+                        np.multiply(x_seg, cell_weight),
+                        axis=(y_index, x_index),
+                    ),
+                    np.sum(cell_weight),
+                    out=output_seg,
+                )
```

Comment:
The `nansum` is taking the most time. Wrapping in Cython might improve performance (but trying to avoid this).

Keeping data in Xarray and using Xarray APIs _might_ be performant.

Reply:
This looks resolved in the latest commit.